src/clients/HttpTunneler.cc

```diff
@@ -223,6 +231,7 @@ Http::Tunneler::handleReadyRead(const CommIoCbParams &io)
         const auto error = new ErrorState(ERR_READ_ERROR, Http::scBadGateway, request.getRaw(), al);
         error->xerrno = rd.xerrno;
         bailWith(error);
+        connection->close();
```

Comment:
Should this connection closure be moved inside bailWith(error)?

Reply:
> We need to refactor Tunneler code so that the connection closure and timeout handlers are clearly always in sync with the connection. Tunneler::callBack() removes the handlers (while keeping the connection open). I suspect it should be _moved_ into a new disconnect() method sketched below. After refactoring, you will have:
> 
>     1. bailWith() calls callBack(), disconnect(true).
>     2. sendSuccess() calls callBack(), disconnect(false).
>     3. swanSong() calls sendSuccess() or bailWith() if needed. Otherwise, calls disconnect(false).
>     4. callBack() becomes a low-level helper method that does not worry about `connection` at all.
>     5. disconnect() is a new low-level helper method that resets connection-related info:
> 

The scheme is looking good to me, and I have already implement it localy. However probably we should not always close the server connection on Peer error responses  in the case of Http::Tunneler (in the PeerConnector case there is not such issue).

Inside the FwdState::tunnelEstablishmentDone exist the following TODO:
   >  // TODO: Reuse to-peer connections after a CONNECT error response.

I can not find any reason why we should reuse a connection we seen an error page, but probably it is the CachePeer job to close imediatelly the connection if it believes we should not reuse it.

The above problem can be solved if we use a parameter in bailWith which means fatal or non fatal error, and close the connection only in fatal case.

Reply:
> I can not find any reason why we should reuse a connection we seen an error page, 

Just because the peer did not like our CONNECT request, does not mean we cannot reuse the connection to that peer for other (future) requests. The connection may still be persistent, and we should reuse persistent connections.

> probably it is the CachePeer job to close imediatelly the connection if it believes we should not reuse it.

Yes, probably. And, after we teach Squid to copy connection details (for logging, etc.), CachePeer can place the reusable to-server connection into the pconn pool after non-closing errors as well. Until then, the caller has to do all that. All that is outside this PR scope though.

> probably we should not always close the server connection on Peer error responses in the case of Http::Tunneler

Agreed. And it is Tunneler who should determine whether the connection is still reusable. However, implementing that logic is outside this PR IMO. We can always close in bailWith() for now AFAICT.

BTW, always closing in bailWith() may fix Tunneler::handleTimeout() that does not close the connection right now directly or indirectly (but probably should close it, indirectly via bailWith()).

> The above problem can be solved if we use a parameter in bailWith

I do not think we have to make things more complex in this PR. We can leave that old TODO unaddressed here, right?

Reply:
commit 7ea5e844 implements your suggestion and should be ok.